src/client/debugger/extension/adapter/logging.ts

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+
+import * as fs from 'fs';
+import { inject, injectable } from 'inversify';
+import * as path from 'path';
+import { DebugAdapterTracker, DebugAdapterTrackerFactory, DebugConfiguration, DebugSession, ProviderResult } from 'vscode';
+import { DebugProtocol } from 'vscode-debugprotocol';
+
+import { IFileSystem } from '../../../common/platform/types';
+import { StopWatch } from '../../../common/utils/stopWatch';
+import { EXTENSION_ROOT_DIR } from '../../../constants';
+
+class DebugSessionLoggingTracker implements DebugAdapterTracker {
+    private readonly session: DebugSession;
+    private readonly enabled: boolean = false;
+    private stream: fs.WriteStream | undefined;
+    private timer = new StopWatch();
+
+    constructor(session: DebugSession, fileSystem: IFileSystem) {
+        this.session = session;
+        this.enabled = this.session.configuration.logToFile as boolean;
+        if (this.enabled) {
+            const fileName = `debugger.vscode_${this.session.id}.log`;
+            this.stream = fileSystem.createWriteStream(path.join(EXTENSION_ROOT_DIR, fileName));
+        }
+    }
+
+    public onWillStartSession() {
+        this.timer.reset();
+        this.log(`Starting Session:\n${this.stringify(this.session.configuration)}\n`);
+    }
+
+    public onWillReceiveMessage(message: DebugProtocol.Message) {
+        this.log(`Client <-- Adapter:\n${this.stringify(message)}\n`);
+    }
+
+    public onDidSendMessage(message: DebugProtocol.Message) {
+        this.log(`Client --> Adapter:\n${this.stringify(message)}\n`);
+    }
+
+    public onWillStopSession() {
+        this.log(`Stopping Session\n`);
```

Comment:
Personal opinion so feel free to disregard: you don't need to use a template literal here here since you're not building a complex string (multiline strings or expression interpolation)

```suggestion
        this.log('Stopping Session\n');
```